src/docs/arch-concurrency.dox

```diff
@@ -0,0 +1,107 @@
+/*! @arch_page arch-concurrency WiredTiger concurrency management
+
+WiredTiger is a multi-threaded application and needs to manage concurrent access to shared memory
+locations. Many mechanisms are defined in WiredTiger to safely access shared memory locations. This
+page describes what those mechanisms are and the intended usage of them in the code base.
+
+@section volatile Volatile
+
+A controversial tool in the C toolkit. WiredTiger utilizes volatile for one reason, managing
+compiler optimization. Leveraging the definitions from "Is Parallel Programming Hard, And, If So,
+What Can You Do About It?", Paul E. McKenney, WiredTiger utilizes volatile to prevent the
+following type of compiler optimizations: Load fusing, store fusing, invented loads, invented
+stores and code reordering. WiredTiger defines the following macros WT_READ_ONCE and
+WT_WRITE_ONCE, these macros, if used correctly, prevent the described forms of compiler
+optimization. Additionally some variables in WiredTiger are defined with the volatile keyword if
+it is known that they will be frequently accessed in a concurrent context.
+
+@section locking Locking
+
+There are three types of locks in WiredTiger:
+ - Condition Variables
+ - Spin Locks
+ - Read Write (RW) Locks
+Each mechanism guarantees that once a thread relinquishes its lock and another thread obtains the
+same lock, all modifications made by the first thread will be visible to the thread now holding
+the lock.
+
+There are two implementations of condition variables in WiredTiger respectively for POSIX and MSVC
+under the same interface. The specific implementation is determined at compile time based on the
+platform.
+
+There are three different implementations for spin locks in WiredTiger under the same interface:
+ - An implementation using GCC's builtin atomic operations.
+ - An implementation using a POSIX based mutex.
+ - An implementation for MSVC.
```

Comment:
Two of the 3 are not true spinlocks. @ajmorton has already mentioned the relevant ticket WT-12960.

Reply:
Removed per this conversation: https://github.com/wiredtiger/wiredtiger/pull/10575#discussion_r1612519348